File: hangman_py/RandomWords.py
import os
import sys
import random

if len(sys.argv) < 2:
  print >> sys.stderr, "A number of random words is expected but missing."
  sys.exit(1)

DICT_FILE = os.environ.get('hangman_dict', 'words.txt')

dict = []
try:
  for line in open(DICT_FILE, 'rb'):
    word = line.strip()
    if len(word) > 0:
      dict.append(word.upper())

except IOError:
  print >> sys.stderr, "Cannot open dictionary file '%s' for reading!" % DICT_FILE
  sys.exit(1)

try:
  count = int(sys.argv[1])
except TypeError:
  count = 0

if count < 1 or count > len(dict):
  print >> sys.stderr, "%s is not a number, or its value is out of [1, %d]" % (sys.argv[1], len(dict))
  sys.exit(1)

out = set()
size = len(dict)
random.seed()

def rand():
  return int(random.random() * size)

i = count
while i >= 1:
  idx = (rand() * 10 + rand() + int(rand() / (rand() + 1))) % size
  if idx not in out:
    out.add(idx)
    print dict[idx].upper()
    i -= 1

File: hangman_ruby/RandomWords.rb
require 'set'

if ARGV.size < 1 then
  $stderr << "A number of random words is expected but missing.\n"
  exit(false)
end

DICT_FILE = ENV['hangman_dict'] || 'words.txt'

dict = []
begin
  File.open(DICT_FILE, 'rb') do |f|
    f.each_line do |line|
      word = line.chomp
      dict << word if word.size > 0
    end
  end
rescue SystemCallError
  $stderr << "Cannot open dictionary file '%s' for reading!\n" % DICT_FILE
  exit(false)
end

count = nil
if (count = ARGV[0].to_i) < 1 || count > dict.size then
  $stderr << "%s is not a number, or its value is out of [1, %d]\n" % [ARGV[0], dict.size]
  exit(false)
end

out = Set.new
size = dict.size
r = Random.new(Time.now.to_i)
i = count
while i >= 1 do
  idx = (r.rand(size) * 10 + r.rand(size) + Integer(r.rand(size) / (r.rand(size) + 1))) % size
  if out.add?(idx) then
    $stdout << dict[idx].upcase << "\n"
    i -= 1
  end
end
